### PR TITLE
Multi arch guestbook example

### PR DESCRIPTION
**Pull request by mkumatag**
Guestbook example is part of [Conformance tests](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubectl/kubectl.go#L354). So this PR is for converting this example multi architecture to support conformance tests on all possible architectures.

### CODE REVIEW

**Comment by ahmetb** [comment]
Thanks  @mkumatag. I'm concerned about 2 parts:

1. A beginner coming to this repo for examples material will no longer be able to `docker build -t guestbook .`
2. As far as I can tell, the output image will be in format `gcr.io/google-samples/guestbook:v5-amd64`. Can we keep it as `:v5` for amd64 specifically not to break/complicate the tutorials?

**Comment by mkumatag** [comment]
> Thanks @mkumatag. I'm concerned about following two points:
> 
> 1.  A beginner coming to this repo for examples material will no longer be able to docker build -t guestbook .

IMO it should be okay if we update README file to use makefile to build images. The reason behind having one Dockefile is to avoid any extra burden in the future. 
> 2. As far as I can tell, the output image will be in format gcr.io/google-samples/guestbook:v5-amd64. Can we keep it as :v5 for amd64 specifically, not to break/complicate the tutorials?

This is taken care, the images will be in format of - `gcr.io/google-samples/gb-frontend-<ARCH>:v5` and I'm maintaining backward compatibility [here](https://github.com/mkumatag/examples/blob/gb_multiarch/guestbook/php-redis/Makefile#L78) and this will tag image like - `gcr.io/google-samples/gb-frontend:v5` for `amd64` ARCH.

File: guestbook/php-redis/Dockerfile [hunk 1]
```diff
@@ -12,7 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM php:5-apache
+FROM BASEIMAGE
```
Comment (sebgoa) [inline guestbook/php-redis/Dockerfile]: What is this BASEIMAGE ? never seen this before

File: guestbook/php-redis/Makefile [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# Copyright 2017 The Kubernetes Authors. All rights reserved.
```
Comment (sebgoa) [inline guestbook/php-redis/Makefile]: you add a Makefile but there is no reference to it anywhere. Not in the README. So how is the user supposed to know what to do with this Makefile ?

File: guestbook/php-redis/Makefile [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+IMAGE = $(REGISTRY)/gb-frontend
+MULTI_ARCH_IMG = $(IMAGE)-$(ARCH)
+
+# Set default base image dynamically for each arch
+ifeq ($(ARCH),amd64)
+	BASEIMAGE=php:5-apache
+endif
+ifeq ($(ARCH),arm)
+	QEMUARCH=arm
+	BASEIMAGE=armhf/php:5.6-apache
```
Comment (sebgoa) [inline guestbook/php-redis/Makefile]: These are not official images. Should we rely on them ?

File: guestbook/php-redis/Dockerfile (hunk 1 above)
Comment (sebgoa) [inline guestbook/php-redis/Dockerfile]: I see that you sed this in the Makefile. I just think it is going to be confusing for the first time user. Plus there is no reference to your Makefile anywhere.

**Comment by sebgoa** [comment]
I suggest we create a separate examples under `/staging` which is a multi-arch version of guestbook.

File: guestbook/php-redis/Makefile (hunk 2 above)
Comment (mkumatag) [inline guestbook/php-redis/Makefile]: np, I'll add the information in README.md file

File: guestbook/php-redis/Makefile (hunk 3 above)
Comment (mkumatag) [inline guestbook/php-redis/Makefile]: These are official images, here is the information - https://github.com/docker-library/official-images#architectures-other-than-amd64

**Comment by mkumatag** [comment]
> I suggest we create a separate examples under /staging which is a multi-arch version of guestbook.

Do you want me to copy or move??

File: guestbook/php-redis/Dockerfile (hunk 1 above)
Comment (ahmetb) [inline guestbook/php-redis/Dockerfile]: I'm thinking just to eliminate confusion, maybe we should maintain 2 copies: Dockerfile and Dockerfile.multiarch?

Comment (mkumatag) [inline guestbook/php-redis/Dockerfile]: The problem with that is additional maintenance.. BTW all the imaages including infra in `kubernetes/kubernetes` following this approach. @luxas any comments?

File: guestbook/php-redis/Makefile (hunk 3 above)
Comment (sebgoa) [inline guestbook/php-redis/Makefile]: ok thanks

**Comment by mkumatag** [comment]
@sebgoa I don't really understand the reason for copying the same code under `/staging` directory. Any specific reason for that?

**Comment by sebgoa** [comment]
user experience.

By adding a makefile and supporting multiple arch, you are forcing the new comer to read the Makefile and understand what is happening. This seems obvious to most people, but for newcomers we should aim for the shortest path to running the example.

These examples are not only used by conformance tests but also in the documentation.

So my suggestion would be to keep a guestbook example for docs, and put your multi-arch support (for conformance tests) in a separate directory.

File: guestbook/php-redis/Dockerfile (hunk 1 above)
Comment (luxas) [inline guestbook/php-redis/Dockerfile]: @sebgoa pretty much every other Dockerfile in the k/k, k/contrib and other repos use this standardized pattern. I think adding a comment above way better than maintaining many copies.

File: guestbook/php-redis/Makefile (hunk 3 above)
Comment (luxas) [inline guestbook/php-redis/Makefile]: @mkumatag armhf is deprecated in favor for `arm(32|64)v(5|6|7|8)`

File: guestbook/php-redis/Dockerfile [hunk 4]
```diff
@@ -12,7 +12,9 @@
...
+
```
Comment (luxas) [inline guestbook/php-redis/Dockerfile]: Please also copy the comment I have made for this statement

**Comment by ahmetb** [comment]
I think as long as the Dockerfile is directly buildable by beginner users, I'm not too worried about googlers pushing the image (=me). That's why I initially suggested having 2 dockerfiles in the same dir. Different dir works too, just makes it harder to maintain in both places if they end up in different directories.

**Comment by mkumatag** [comment]
What if I modify my Makefile slightly different to accommodate not to change Dockerfile, sed will look something like this : `sed "s|FROM php.*|FROM ${BASEIMAGE}|g" Dockerfile`. With this Dockerfile remains intact, and only worry about keeping FROM command unchanged. If changed then need to modify Makefile as well. WDYT on this?

**Comment by ahmetb** [comment]
@mkumatag sounds OK to me. @sebgoa ?

**Comment by luxas** [comment]
@mkumatag That sounds good to me as well. Please document in the README that a normal user can just do `docker build -t foo .`, but someone releasing the image should use `make foo bar` instead

**Comment by mkumatag** [comment]
@sebgoa @ahmetb did a required changes, PTAL at latest commit.

**Comment by sebgoa** [comment]
no changes in the README and the redis-master is untouched, is that on purpose ?

**Comment by mkumatag** [comment]
I don't know which redis-master are you talking about, I don't find that folder anywhere in this example. For README file I need some more time, that I'll try to submit part of another PR.

**Comment by sebgoa** [comment]
this example needs a redis-master. It is defined in the redis-master deployment files. The image is `gcr.io/google_containers/redis:e2e`

so if we update the redis-slave and frontend, we should also update the master to run on multi-arch otherwise it won't work...

plus the image names most likely need to be updated in the manifests

just my 2cts

**Comment by mkumatag** [comment]
Yes you are right.! redis master image already taken care and ported for multi arch and present under - https://github.com/kubernetes/kubernetes/tree/master/test/images/redis.

And about manifest files - Currently tests are using manifests from https://github.com/kubernetes/kubernetes/tree/master/examples/guestbook location which will not be valid in the future, so my plan is to convert these manifest files to Jinja2 template and replace the imagenames runtime.

**Comment by mkumatag** [comment]
@sebgoa do you have any other review comments?

**Comment by luxas** [comment]
ping @sebgoa @ixdy @ahmetb

File: guestbook/php-redis/Makefile (hunk 2 above)
Comment (sebgoa) [inline guestbook/php-redis/Makefile]: I don't see any changes to the Makefile ?

Comment (mkumatag) [inline guestbook/php-redis/Makefile]: @sebgoa you mean README.md file?

**Comment by sebgoa** [comment]
so you have conformance tests that use manifests from the main repo, two images being built form this very repo and one image built from a test directory in the main repo ?

IMHO, the conformance tests should use manifests in this repo (so we can finish cleaning up the main repo) and *all* images should be built from the same place.

but I am fine merging to remove the blocker if I see a mention in the README about using `make` to do different arch builds.

File: guestbook/php-redis/Makefile (hunk 2 above)
Comment (sebgoa) [inline guestbook/php-redis/Makefile]: yes sorry...the README.md

**Comment by mkumatag** [comment]
> so you have conformance tests that use manifests from the main repo, two images being built form this very repo and one image built from a test directory in the main repo ?

Yes thats true.
> 
> IMHO, the conformance tests should use manifests in this repo (so we can finish cleaning up the main repo) and all images should be built from the same place.
> 

This is very difficult at the moment wrt how test infra is designed to pick up manifest files and use in the tests.
> but I am fine merging to remove the blocker if I see a mention in the README about using make to do different arch builds.

Will modify the README.md file with instructions.

File: guestbook/php-redis/Makefile (hunk 2 above)
Comment (mkumatag) [inline guestbook/php-redis/Makefile]: @sebgoa added instruction in README.md file, PTAL

File: guestbook/README.md [hunk 5]
```diff
@@ -55,6 +56,31 @@ If you see a url response, you are ready to go. If not, read the [Getting Starte
 
 All the files referenced in this example can be downloaded [from GitHub](https://git.k8s.io/examples/guestbook).
 
+### Building The Docker Images
```
Comment (ahmetb) [inline guestbook/README.md]: Looks like there is repetition with this heading and in the sentence below.

You can just change the capitalization to 

     ### Building the Docker images

and remove the following sentence: `Building docker images:`.

Comment (mkumatag) [inline guestbook/README.md]: fixed your comment in the latest commit, PTAL

**Comment by sebgoa** [comment]
ok merging, thanks

**Comment by mkumatag** [comment]
@sebgoa @ahmetb @ixdy can someone push images to google_samples repo by following instructions at https://github.com/kubernetes/examples/tree/master/guestbook#building-the-docker-images

**Comment by mkumatag** [comment]
Ping @ixdy

**Comment by ixdy** [comment]
Sorry for the delay. I don't have access to gcr.io/google-samples - @ahmetb can you assist?

**Comment by ahmetb** [comment]
Sorry for the delay, I just pushed the following images:

```
➜  ~ docker images | grep gb-
gcr.io/google-samples/gb-frontend-ppc64le     v5                                         de4b3d45c7ad        59 seconds ago      406MB
gcr.io/google-samples/gb-redisslave-ppc64le   v2                                         72606f95f73f        3 minutes ago       105MB
gcr.io/google-samples/gb-frontend-arm64       v5                                         f907eec7c4b3        4 minutes ago       365MB
gcr.io/google-samples/gb-redisslave-arm64     v2                                         64e75b265263        4 minutes ago       94.1MB
gcr.io/google-samples/gb-redisslave-arm       v2                                         d48c0b56be51        4 minutes ago       15.8MB
gcr.io/google-samples/gb-redisslave-amd64     v2                                         23e31172abbc        5 minutes ago       98.9MB
gcr.io/google-samples/gb-redisslave           v2                                         23e31172abbc        5 minutes ago       98.9MB
gcr.io/google-samples/gb-frontend-arm         v5                                         d6a5be0ed52a        6 minutes ago       319MB
gcr.io/google-samples/gb-frontend-amd64       v5                                         524e6104887a        9 minutes ago       405MB
gcr.io/google-samples/gb-frontend             v5                                         524e6104887a        9 minutes ago       405MB
```

If you think the tutorials need updating, those now live in the docs (kubernetes/kubernetes.github.io) repo now but I suspect that is the case.

Out of curiosity: @mkumatag do these multi-arch images have to be separate image names? Can't we name these like `gcr.io/google-samples/gb-redisslave:v2-arm64`, instead of `gcr.io/google-samples/gb-redisslave-arm64:v2`? Is there a limitation or convention there?

**Comment by mkumatag** [comment]
I think it's all about the convention we are using in the kubernetes from day 1. @luxas please let us know if there any specific reason behind naming convention for these images.

**Comment by luxas** [comment]
We had long-ish discussions on this, and decided that `registry/binary-arch:version` is the best way for now (>1year ago). However, once gcr.io starts supporting [Docker Registry API v2 schema 2](https://docs.docker.com/registry/spec/manifest-v2-2/) (which has been an outstanding request for more than a year, see https://www.youtube.com/watch?v=ZdzKQwMjg2w)

This should really not be visible to users, but as we use gcr.io we're stuck here... :(
However, I've heard reports that the feature is in the works now, so probably in time for v1.9 :smile:

**Pull request merged**